src/librustc_mir/util/pretty.rs

```diff
@@ -346,11 +339,93 @@ where
         comment(tcx, data.terminator().source_info),
         A = ALIGN,
     )?;
+
+    write_extra(tcx, w, |visitor| {
+        visitor.visit_terminator(current_location.block, data.terminator(), current_location);
+    })?;
+
     extra_data(PassWhere::AfterLocation(current_location), w)?;
 
     writeln!(w, "{}}}", INDENT)
 }
 
+/// After we print the main statement, we sometimes dump extra
+/// information. There's often a lot of little things "nuzzled up" in
+/// a statement.
+fn write_extra<'cx, 'gcx, 'tcx, F>(
+    tcx: TyCtxt<'cx, 'gcx, 'tcx>,
+    write: &mut Write,
+    mut visit_op: F,
+) -> io::Result<()>
+where F: FnMut(&mut ExtraComments<'cx, 'gcx, 'tcx>)
+{
+    let mut extra_comments = ExtraComments {
+        _tcx: tcx,
+        comments: vec![],
+    };
+    visit_op(&mut extra_comments);
+    for comment in extra_comments.comments {
+        writeln!(write, "{:A$} // {}", "", comment, A = ALIGN)?;
+    }
+    Ok(())
+}
+
+struct ExtraComments<'cx, 'gcx: 'tcx, 'tcx: 'cx> {
+    _tcx: TyCtxt<'cx, 'gcx, 'tcx>, // don't need it now, but bet we will soon
```

Comment:
A potential alternative: Instead of having `ExtraComments` carry a `tcx` and get passed into a callback closure, you might consider having the closure itself be responsible for carrying the `tcx`?

At least, that's what I ended up deciding to do as part of a cleanup of the drop_flag_effects code, see [here](https://github.com/rust-lang/rust/pull/46537/commits/71e619f47800298c584a6315a2b8b584aaf27282#diff-1241cb461608f1c7ff8856111634ffcaR116), where I found it more pleasant to put one or two type params in the signature than to pass around `TyCtxt` and its three required lifetimes...

Reply:
In this case, it's a mir visitor, not a closure.

Reply:
But also, I don't personally mind passing around the three lifetimes. Especially once the in-band lifetimes work lands makes its way into beta (it's in nightly already...), that will be quite lightweight.